drift.py

```diff
@@ -1,140 +1,77 @@
-        return tags
-
-    def tag_variants(self, concrete_spec, spec, name):
-        tags = []
-        if "parameters" in concrete_spec[0][name]:
-            # Check for variants added or modified by comparing new spec to old.
-            for param, value in concrete_spec[0][name]["parameters"].items():
-                if param not in self.last[spec][0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-added"]
-                    else:
-                        tags += ["variant-added"]
-                elif value != self.last[spec][0][name]["parameters"][param]:
-                    if param == "patches":
-                        tags += ["patches-modified"]
-                    else:
-                        tags += ["variant-modified"]
-            # Compare old to new to see if any variants were removed.
-            for param, value in self.last[spec][0][name]["parameters"].items():
-                if param not in concrete_spec[0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-removed"]
-                    else:
-                        tags += ["variant-removed"]
-            return tags
+        for abstract_spec in self.specs:
+            # Separate out spec version and name.
+            verToken = abstract_spec.find("@")
+            if verToken < 0: verToken = len(abstract_spec)
+            out = run(f"spack versions --safe-only {abstract_spec}")
+            # Don't attempt to concretize if the version doesn't yet exist.
+            if abstract_spec[verToken+1:] in out.stdout:
+                out = run(f"spack spec --yaml {abstract_spec}")
+                # If concretization successful check the resulting concrete specs.
+                if out.returncode == SUCCESS:
+                    concrete_spec = spack.spec.Spec().from_yaml(out.stdout)
+                    if abstract_spec in self.last and self.last[abstract_spec] != concrete_spec:
+                        # Construct Result
+                        result = Result(
+                            abstract_spec[:verToken],
+                            abstract_spec[verToken+1:],
+                            commit.hash,
+                            # Waiting for spack diff to complete tags.
+                            [],
+                            str(commit.author_date))
+                        # Send result to drift-server
+                        send(result)
+                    # Save concrete spec as last spec
+                    self.last[abstract_spec] = concrete_spec
+                # Mark failing concretization points.
+                else:
+                    # Construct Result
+                        result = Result(
+                            abstract_spec[:verToken],
+                            abstract_spec[verToken+1:],
+                            commit.hash,
+                            ["concretization-failed"],
```

Comment:
If the only difference between this and the above is the tag,  I would set the tags in the two sections, and then just have this Result creation once.

Reply:
Right, so this is a little tricky I think because we have two cases where we want to send results to the `drift-server`. We either want to send a result if the package doesn't concretize, or concretizes and is different than the previous version. But we don't want to send anything if the package concretizes and is the same as before. So I think to put the result creation in a central place we'd have to duplicate the previous checks again which I thought was kinda ugly. However, I agree this one is also not great haha. What do you think?

Reply:
I still think there is redundancy. You can just do:
```python
if case1:
    tags = # some result from spack diff
else:
    tags = ["concretization-failed"]

result = Result(abstract_spec[:verToken],
                         abstract_spec[verToken+1:],
                         commit.hash,
                         tags,
                        [],
                        str(commit.author_date)
)
send(result)
```

Reply:
Gotcha I think the issue is that we'd have to skip the other case in the inner if.
```python
if outer_case:
    if inner_case:
        tags = # some result from spack diff
    else:
        # this case should exit and not continue to the send stage
else:
    tags = ["concretization-failed"]

result = Result(abstract_spec[:verToken],
                         abstract_spec[verToken+1:],
                         commit.hash,
                         tags,
                        [],
                        str(commit.author_date)
)
send(result)
```

Reply:
If that's in a loop you can use a continue, I'd do like:
```python
if outer_case:
    if not inner_case:
        continue    
    tags = # some result from spack diff
else:
    tags = ["concretization-failed"]

result = Result(abstract_spec[:verToken],
                         abstract_spec[verToken+1:],
                         commit.hash,
                         tags,
                        [],
                        str(commit.author_date)
)
send(result)
```

Reply:
The issue I was trying to get around with this suggestion is having the result defined twice, so if you make a change you need to update both (which could lead to errors, assuming not every developer is perfect!)